### PR TITLE
[Tensor Parallel] Add _parallelize_linear

### PR DESCRIPTION
**Pull request by XilunWu**
- added low-level parallelize function for `nn.Linear`
- added unit tests and test helper to further remove existing repeated test code in future PR

### CODE REVIEW

File: test/spmd/tensor/parallel/test_tp_examples.py [hunk 1]
```diff
@@ -511,6 +515,64 @@ def test_self_attn_replacement_megatron_e2e(self):
...
+        self.assertEqual(
+            distribute_tensor(
+                model.bias, device_mesh, [Replicate()]
+            ).to_local(),
+            model_tp.bias.to_local(),
+        )
+
+        output = model(inp)
+        output_tp = model_tp(inp)
+        self.assertEqual(output, output_tp.to_local())
```
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_tp_examples.py:547]: Can we also add backward and calling optimizer? Thanks!

File: spmd/tensor/parallel/api.py [hunk 2]
```diff
@@ -86,3 +91,93 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+            distribute_tensor(module.bias, device_mesh, bias_placements)
+        )
+        module.register_parameter("bias", bias)
+    # input
+    if parallel_style._prepare_input is not None:
+        input_fn: Callable[
+            [Sequence[Union[torch.Tensor, DTensor]], Tuple[DTensor, ...]]
+        ] = lambda inputs: tuple(
+            parallel_style._prepare_input(tensor, device_mesh)
+            for tensor in inputs
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:175]: I think we can just focus on inputs[0] for now and we can use `distribute_module` API.

**Review by fduwjj — COMMENTED** [review]
Thanks for working on this and I have some comments about test cases and actual logic. LMK if you have more questions.

File: spmd/tensor/parallel/api.py (hunk 2 above)
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:175]: Will address it later.

File: spmd/tensor/parallel/style.py [hunk 3]
```diff
@@ -28,14 +28,34 @@ class ParallelStyle(ABC):
...
+
+
+class ColwiseParallel(ParallelStyle):
+    """
+    Partitioning the column of a tensor or module.
+    We assume the input to be a replicated :class:``DTensor`` and output to be a sharded :class:``DTensor``.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(make_input_replicate_1d, None)
```
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:48]: As per discussion with Wanchao, we believe we can use make_output_replicate_1d for prepare_out as well. So we removed Optional for both prepare_input and prepare_output.

**Comment by fduwjj** [comment]
Also you might want to rebase on top of https://github.com/pytorch/tau/pull/623. Thanks!

**Comment by XilunWu** [comment]
Signals are all green. Ready for review.

File: spmd/tensor/parallel/__init__.py [hunk 4]
```diff
@@ -11,9 +11,9 @@
 
 from spmd.tensor.parallel.style import (
     ParallelStyle,
-    PairwiseParallel,
     RowwiseParallel,
     ColwiseParallel,
+    PairwiseParallel,
```
Comment (fduwjj) [inline spmd/tensor/parallel/__init__.py:16]: OK, if you really want to sort it, can we sort it alphabetically?

File: spmd/tensor/parallel/api.py [hunk 5]
```diff
@@ -12,7 +12,12 @@
     Placement,
 )
 from spmd.tensor.parallel import TensorParallelMultiheadAttention
-from spmd.tensor.parallel.style import ParallelStyle, PairwiseParallel
+from spmd.tensor.parallel.style import (
+    ParallelStyle,
+    RowwiseParallel,
+    ColwiseParallel,
+    PairwiseParallel,
+)
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:20]: ditto

File: spmd/tensor/parallel/api.py [hunk 6]
```diff
@@ -89,6 +94,92 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+            The dimension of :class:``DeviceMesh`` on which we perform
+            Tensor Parallelism.
+            Default: 0
+    Returns:
+        None
+    """
+
+    if not isinstance(module, nn.Linear):
+        raise RuntimeError(
+            f"Tensor parallel module expects a torch.nn.Linear module but received {type(module)}!"
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:152]: I got confusing by this saying "Tensor parallel module". Can we just "Expect ..."?

File: spmd/tensor/parallel/api.py [hunk 7]
```diff
@@ -89,6 +94,92 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
                 module.register_module(n, tp_multi_head_attention)
 
 
+# Define partition functions needed to parallelize Linear modules
+def _rowwise_parallelize_fn(name, module, device_mesh):  # pyre-ignore[2, 3]
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:98]: let's probably add linear here. Because for other types of module, rowwise and colwise will be different.

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 8]
```diff
@@ -126,6 +134,161 @@ def __init__(self) -> None:
...
+        inp = torch.rand(*inp_size, device=self.device_type)
+        local_inp = make_input_replicate_1d(
+            DTensor.from_local(inp, device_mesh, [Shard(0)])
+        ).to_local()
+        output = model(local_inp)
+        output_tp = model_tp(inp)
+        self.assertEqual(output, output_tp.to_local())
+
+    @with_comms
+    def test_linear_col_wise_parallel(self):
```
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_parallelize_api.py:220]: Can we consolidate the two test a little bit, the code is way too long and there are multiple lines of duplication. Thanks! Feel free to refactor the whole test file to make it cleaner.

File: spmd/tensor/parallel/api.py [hunk 9]
```diff
@@ -89,6 +94,92 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+    """
+
+    if not isinstance(module, nn.Linear):
+        raise RuntimeError(
+            f"Tensor parallel module expects a torch.nn.Linear module but received {type(module)}!"
+        )
+
+    if not isinstance(parallel_style, ParallelStyle):
+        raise RuntimeError(
+            f"Tensor parallel module expects a ParallelStyle object but received {type(parallel_style)}!"
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:157]: ditto.

**Review by fduwjj — COMMENTED** [review]
Thanks and leave some comments.

File: spmd/tensor/parallel/api.py [hunk 10]
```diff
@@ -89,6 +94,92 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+    The module will be parallelized over the n-d :class:``DeviceMesh``
+    based on the :class:``ParallelStyle``.
+
+    Args:
+        module (nn.Module):
+            The :class:``nn.Module`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            The n-d :class:``DeviceMesh`` over which the `module` will be parallelized.
+            If the mesh is more than 1-dimensional, we convert it to a 1-d :class:``DeviceMesh``
+            by `tp_mesh_dim`.
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:134]: You might want to say something like this:

:class:``DeviceMesh`` object which describes the mesh topology
            of devices for the DTensor.

Let's mention topology.

File: spmd/tensor/parallel/api.py [hunk 11]
```diff
@@ -89,6 +94,92 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+        parallel_style (ParallelStyle):
+            :class:``ParallelStyle`` describes how weight and bias should be distributed
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:136]: Just say Linear

File: spmd/tensor/parallel/api.py (hunk 7 above)
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:98]: To be clear, previous this is inner function inside `_parallize_mlp` so that we don't need to add linear in the name. Also please do add in/o type and comments for this function. Thanks!

Comment (XilunWu) [reply spmd/tensor/parallel/api.py:98]: Thx for catching it!

File: spmd/tensor/parallel/api.py (hunk 10 above)
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:134]: Shall we also mention that n-d mesh will be converted to 1-d mesh as well?

File: spmd/tensor/parallel/api.py (hunk 11 above)
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:136]: Do you mean "weight and bias" -> "the nn.Linear module"

Comment (fduwjj) [reply spmd/tensor/parallel/api.py:136]: yep.

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 8 above)
Comment (XilunWu) [reply test/spmd/tensor/parallel/test_parallelize_api.py:218]: done. TODO: see if existing test code in `test_tp_example.py` can be replaced as well.

File: spmd/tensor/parallel/api.py [hunk 12]
```diff
@@ -181,8 +276,8 @@ def _colwise_parallelize_fn(name, module, device_mesh):  # pyre-ignore[2, 3]
             distribute_module(
                 m,
                 device_mesh,
-                _colwise_parallelize_fn,
-                input_fn=parallel_style._prepare_input  # type: ignore[arg-type, misc] # pyre-ignore[6]
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:185]: You should keep the order as it is, otherwise, mypy might fail.

Comment (XilunWu) [reply spmd/tensor/parallel/api.py:185]: Must be overriding your CI fix by mistake when resolving merge conflict. Thx for pointing out!

File: spmd/tensor/parallel/api.py [hunk 13]
```diff
@@ -89,6 +94,114 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+) -> None:
+    """
+    This function parallelizes the input :class:``nn.Linear`` module in :class:``RowwiseParallel`` style.
+
+    Args:
+        name (str): name of the input module.
+        module (nn.Module): the :class:``nn.Linear`` object to be parallelized.
+        device_mesh (DeviceMesh): :class:``DeviceMesh`` object which describes the mesh topology
+            of devices for the DTensor.
+    Returns:
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:109]: Nit: one new blank line.

File: spmd/tensor/parallel/api.py [hunk 14]
```diff
@@ -89,6 +94,114 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+) -> None:
+    """
+    This function parallelizes the input :class:``nn.Linear`` module in :class:``ColwiseParallel`` style.
+
+    Args:
+        name (str): name of the input module.
+        module (nn.Module): the :class:``nn.Linear`` object to be parallelized.
+        device_mesh (DeviceMesh): :class:``DeviceMesh`` object which describes the mesh topology
+            of devices for the DTensor.
+    Returns:
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:133]: Nit: one new blank line.

File: spmd/tensor/parallel/api.py [hunk 15]
```diff
@@ -89,6 +94,114 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+        parallel_style (ParallelStyle): :class:``ParallelStyle`` describes how the
+            :class:``nn.Linear`` module should be distributed over :class:``DeviceMesh``
+            and how the input and output should be prepared for Tensor Parallelism.
+            :class:``RowwiseStyle``: weight is sharded on dim 1 and bias is replicated.
+            :class:``ColwiseStyle``: weight and bias are both sharded on dim 0.
+            Default: :class:``ColwiseParallel``
+        tp_mesh_dim (int): the dimension of :class:``DeviceMesh`` on which we perform
+            Tensor Parallelism.
+            Default: 0
+    Returns:
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:168]: ditto.

File: spmd/tensor/parallel/api.py [hunk 16]
```diff
@@ -12,7 +12,12 @@
     Placement,
 )
 from spmd.tensor.parallel import TensorParallelMultiheadAttention
-from spmd.tensor.parallel.style import ParallelStyle, PairwiseParallel
+from spmd.tensor.parallel.style import (
+    ParallelStyle,
+    RowwiseParallel,
+    ColwiseParallel,
+    PairwiseParallel,
```
Comment (fduwjj) [inline spmd/tensor/parallel/api.py:19]: Why do you mark the comment as resolved? Looks like this is not sorted? And yes we didn't do that previously and this needs to be corrected. Thanks!

File: spmd/tensor/parallel/style.py [hunk 17]
```diff
@@ -27,21 +27,6 @@ def __init__(self, _prepare_input, _prepare_output) -> None:
         self._prepare_output = _prepare_output  # type: ignore[assignment, misc]
 
 
-class PairwiseParallel(ParallelStyle):
```
Comment (fduwjj) [inline spmd/tensor/parallel/style.py:30]: Can we not move it? This is not related to the change.

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 8 above)
Comment (fduwjj) [reply test/spmd/tensor/parallel/test_parallelize_api.py:223]: Please add TODO in the code if you have not done it, thanks!

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 18]
```diff
@@ -3,9 +3,17 @@
...
 from spmd.testing.common_dtensor import DTensorTestBase, with_comms
-from spmd.tensor import distribute_tensor, DeviceMesh, Shard, Replicate
-from spmd.tensor.parallel import PairwiseParallel, ParallelStyle
-from spmd.tensor.parallel.api import _parallelize_mlp
+from spmd.tensor import distribute_tensor, DeviceMesh, Shard, Replicate, DTensor
+from spmd.tensor.parallel import (
+    ParallelStyle,
+    RowwiseParallel,
+    ColwiseParallel,
+    PairwiseParallel,
```
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_parallelize_api.py:11]: ditto.

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 19]
```diff
@@ -71,6 +79,55 @@ def test_creat_1d_device_mesh_error(self):
...
+        # check backward correctness:
+        #   ensure gradients are same
+        for name, param in local_module.named_parameters():
+            dist_param = dist_module.get_parameter(name)
+            self.assertEqual(
+                param.grad,
+                dist_param.grad.redistribute(
+                    device_mesh=device_mesh, placements=replicate
+                ).to_local(),
+            )
```
Comment (fduwjj) [inline test/spmd/tensor/parallel/test_parallelize_api.py:126]: I believe this can be further consolidated since it looks very similar with L96-L103. Let's probably extract this logic and make it a util func, so that you can make it work for mlp? (For mlp, you can just have a for loop and call this func inside the for loop?

**Review by fduwjj — COMMENTED** [review]
Thanks for the work and making the test/code cleaner. Left some comments and I think the PR does look much better. Just last round of change, we are ready to go.

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 20]
```diff
@@ -131,6 +188,62 @@ def __init__(self) -> None:
...
+        # parallelize model_tp
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        _parallelize_linear(model_tp, device_mesh, rowwise)
+
+        LR = 0.25  # the learning rate we use for testing
+        optim = torch.optim.SGD(model.parameters(), lr=LR)
+        optim_tp = torch.optim.SGD(model_tp.parameters(), lr=LR)
+
+        torch.manual_seed(self.rank)
+        for _ in range(2):
```
Comment (wanchaol) [inline test/spmd/tensor/parallel/test_parallelize_api.py:211]: shouldn't all of the optimizer construction, input generation, be all in the `check_module` util?

File: test/spmd/tensor/parallel/test_parallelize_api.py [hunk 21]
```diff
@@ -131,6 +188,62 @@ def __init__(self) -> None:
...
+            inp = torch.rand(*inp_size, device=self.device_type)
+            local_inp = make_input_replicate_1d(
```
Comment (wanchaol) [inline test/spmd/tensor/parallel/test_parallelize_api.py:213]: why we need to create a DTensor then convert it to a local tensor?

File: spmd/tensor/parallel/api.py (hunk 16 above)
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:19]: just my confusion: do we usually force `import` sorted or leave it in any order?

File: spmd/tensor/parallel/api.py [hunk 22]
```diff
@@ -89,6 +94,114 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+def _parallelize_linear(
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallel_style: ParallelStyle = ColwiseParallel(),
+    tp_mesh_dim: int = 0,
+) -> None:
+    """
+    This function requires that the input module be an object of :class:``nn.Linear``.
+    The module will be parallelized over the n-d :class:``DeviceMesh``
+    based on the :class:``ParallelStyle``.
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:152]: I agree that we change all of them to single quote. I can do it in another PR replacing them. what do you think @fduwjj

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 21 above)
Comment (XilunWu) [reply test/spmd/tensor/parallel/test_parallelize_api.py:213]: equivalent to `local_input = inp.repeat(self.world_size, 1)`. Will replace it.

File: spmd/tensor/parallel/api.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (fduwjj) [reply spmd/tensor/parallel/api.py:152]: Indeed!! Interesting, because I was looking at the example here: https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html. I thought ``(double is for class) and ` is for variable. Is there any detailed rules for the Google DocString so that we can use as a reference?

[the remark this answers is not in the document]
Comment (wanchaol) [reply spmd/tensor/parallel/api.py:152]: @fduwjj https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#docstring-type-formatting PyTorch actually uses google style format, so I guess with `:class:` single quote, with no `:class:`, double quote class, single quote variable?

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 19 above)
Comment (XilunWu) [reply test/spmd/tensor/parallel/test_parallelize_api.py:126]: that makes sense. is `test/spmd/tensor/parallel/utils.py` a good place to put the func?

File: spmd/tensor/parallel/api.py [hunk 23]
```diff
@@ -89,6 +94,114 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
...
+
+def _parallelize_linear(
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallel_style: ParallelStyle = ColwiseParallel(),
+    tp_mesh_dim: int = 0,
+) -> None:
+    """
+    This function requires that the input module be an object of :class:``nn.Linear``.
+    The module will be parallelized over the n-d :class:``DeviceMesh``
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:151]: Thanks for the clarification on `tp_mesh_dim`! 

-> The module will be parallelized over a 1-d :class:`DeviceMesh` based on the :class:`ParallelStyle`.

File: spmd/tensor/parallel/api.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply spmd/tensor/parallel/api.py:152]: Just FYI, 
For requirement "Each line inside a docstrings block must be limited to 80 characters", we can use `black --line-length 80 --preview` to check it.

File: test/spmd/tensor/parallel/test_parallelize_api.py (hunk 19 above)
Comment (XilunWu) [reply test/spmd/tensor/parallel/test_parallelize_api.py]: added TODO.

File: spmd/tensor/parallel/__init__.py [hunk 24]
```diff
@@ -5,8 +5,8 @@
 
 from spmd.tensor.parallel.style import (
     ColwiseParallel,
-    ParallelStyle,
     PairwiseParallel,
+    ParallelStyle,
```
Comment (XilunWu) [inline spmd/tensor/parallel/__init__.py:9]: reorder

File: spmd/tensor/parallel/__init__.py [hunk 25]
```diff
@@ -21,10 +21,10 @@
 
 __all__ = [
     "ColwiseParallel",
-    "TensorParallelMultiheadAttention",
-    "ParallelStyle",
     "PairwiseParallel",
+    "ParallelStyle",
     "RowwiseParallel",
+    "TensorParallelMultiheadAttention",
```
Comment (XilunWu) [inline spmd/tensor/parallel/__init__.py:27]: reorder

File: spmd/tensor/parallel/__init__.py (hunk 24 above)
Comment (fduwjj) [reply spmd/tensor/parallel/__init__.py:9]: Opps, maybe I didn't update my local change... NVM. Thanks for the fix.

**Review by fduwjj — APPROVED** [review]
LGTM. Thanks for working on building the API.
